src/main/scala/is/hail/io/index/IndexReader.scala

```diff
@@ -83,6 +84,88 @@ class IndexReader(hConf: Configuration, path: String, cacheCapacity: Int = 8) ex
     }
   }
 
+  private[io] def binarySearchLowerBound(n: Int, key: Annotation, f: (Int) => Annotation): Int = {
```

Comment:
If you want to reduce a bit of code duplication, both upperBound and lowerBound are one-liners using the following function (which is just what you wrote with one line changed):
```scala
// Returns smallest i, 0 <= i < n, for which p(i) holds, or returns n if p(i) is false for all i.
// Assumes all i for which p(i) is true are greater than all i for which p(i) is false.
// Returned value j has the property that p(i) is false for all i in [0, j),
// and p(i) is true for all i in [j, n).
def findPartitionPoint(n: Int, p: (Int) => Boolean): Int = {
  var left = 0
  var right = n
  while (left < right) {
    val mid = left + (right - left) / 2
    if (p(mid))
      right = mid
    else
      left = mid + 1
  }
  left
}
```
Then
```scala
binarySearchLowerBound(n, key, f) = findPartitionPoint(n, i => ordering.gteq(f(i), key))
```
and
```scala
binarySearchLowerBound(n, key, f) = findPartitionPoint(n, i => ordering.gt(f(i), key))
```
This matches how I think of lowerBound and upperBound nicely: they partition [0, n) into three (possibly empty) intervals, [0, l), [l, u), [u, n), respectively containing all keys less than, equal to, and greater than `key`.

Reply:
That's awesome! Thanks for the comment!